zapio/writer.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapio
+
+import (
+	"bytes"
+	"io"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// Writer is an io.Writer that writes to the provided Zap logger, splitting log
+// messages on line boundaries.
+//
+// Writer must be closed when finished to flush buffered data to the logger.
+type Writer struct {
+	Log   *zap.Logger   // log to write to
+	Level zapcore.Level // log level to write at
+
+	buff bytes.Buffer
+}
+
+var (
+	_ zapcore.WriteSyncer = (*Writer)(nil)
+	_ io.Closer           = (*Writer)(nil)
+)
+
+// Write writes the provided bytes to the underlying logger at the configured
+// log level and returns the length of the bytes.
+func (w *Writer) Write(bs []byte) (n int, err error) {
+	// Skip all checks if the level isn't enabled.
+	if !w.Log.Core().Enabled(w.Level) {
+		return len(bs), nil
+	}
```

Comment:
(no action needed, just an observation) If we used a constructor, we could do this check once.

I don't think this is going to affect performance in a meaningful way so not worth changing to constructors for this reason though.

Reply:
I considered a constructor first but I realized that meant that we need at least two parameters: level and logger. To leave room for parameters in the future, we would have had to use a parameter struct or functional options, and that seems like a lot of surface area for a simple writer.

I figured we can start with the minimal struct-only approach. We can always add a constructor later; it should be easy to continue to maintain the behavior of respecting the Logger and Level fields since any constructor-based variant with any unexported fields will still always need a Logger and a Level.

Reply:
Agree, makes sense to start with the struct